### DISCUSSION TITLE
about c# sdk

### DISCUSSION THREAD

**Discussion by weianweigan** [milvus-io/milvus#18333] [Q&A and General discussion] [answered]
Hi,i am a c# developer , i want to use milvus to search similar 3d model.
I found that there is not a c# sdk for milvus,so i created a [repository](https://github.com/weianweigan/milvus-sdk-csharp) and made a logo for it. I use github's default [Apache-2.0 license](https://github.com/weianweigan/milvus-sdk-csharp/blob/develop/LICENSE). I want to if there is a license problem about my work?
And is there an official plan to develop a c# version sdk?

**Comment by yhmo** [comment, marked as answer]
Nice work!
I think Apache 2.0 license is ok since the Miivus project also use the same license.
Currently we provided four kinds of sdk: Python/Java/Go/Node. Officially we have no plan to develop c# sdk. Some contributors are working on sdk for other language, such as the sdk for Rust: https://github.com/milvus-io/milvus-sdk-rust, the sdk project is under the Milvus group. It is depends on you to decide make it as a personal project or contribute it as a tool of milvus group. And the most important thing is, to make a sdk valuable, there are lot of work to do:
1. documentation, how to use, example
2. bug fix
3. update the sdk when Milvus api changed
4. full testing, aotumation

**Reply by weianweigan** [reply]
Thanks for your reply!
It's a good choice and i am glad that make it as a tool of milvus group if it is possible. And developers who want to use c# can know it.
Up to now,there is only alpha version released.As you say, there are lots of work to do.I am working for a general availability version.

**Reply by yhmo** [reply]
Ok, we have created a new repo for c# sdk under the milvus group: https://github.com/milvus-io/milvus-sdk-csharp
The milvus v2.1 has more interfaces than v2.0, actually lot of work to do. We recommend sdk developers to learn more details from python/java sdk. Feel free to contact me if you have any questions about the sdk implementation.

**Reply by weianweigan** [reply]
Ok, I will fork and move to new repos.

**Reply by yhmo** [reply]
@weianweigan 
I notice you closed your pull rquest: https://github.com/milvus-io/milvus-sdk-csharp/pull/1
Seems the project access authority was not properly set, now I add your name into maintainer list. Please submit your pull request again.

**Reply by weianweigan** [reply]
> @weianweigan I notice you closed your pull rquest: [milvus-io/milvus-sdk-csharp#1](https://github.com/milvus-io/milvus-sdk-csharp/pull/1) Seems the project access authority was not properly set, now I add your name into maintainer list. Please submit your pull request again.

Ok, i will reopen it.Is there any other requirements and can i just merge pr directly now?
And i hava another problem.I found that milvus-2.1.0 has published. But in the [milvus-proto](https://github.com/milvus-io/milvus-proto), the newest version is v2.0.1 . If i want to get v2.1.0 *.proto files, should i waiting until milvus-proto update or use the proto from [milvus/internal/proto](https://github.com/milvus-io/milvus/tree/master/internal/proto)?

**Reply by yhmo** [reply]
> > @weianweigan I notice you closed your pull rquest: [milvus-io/milvus-sdk-csharp#1](https://github.com/milvus-io/milvus-sdk-csharp/pull/1) Seems the project access authority was not properly set, now I add your name into maintainer list. Please submit your pull request again.
> 
> Ok, i will reopen it.Is there any other requirements and can i just merge pr directly now? And i hava another problem.I found that milvus-2.1.0 has published. But in the [milvus-proto](https://github.com/milvus-io/milvus-proto), the newest version is v2.0.1 . If i want to get v2.1.0 *.proto files, should i waiting until milvus-proto update or use the proto from [milvus/internal/proto](https://github.com/milvus-io/milvus/tree/master/internal/proto)?

I add some comments in the pr, please take a look. 
You are the maintainer of the project, so you have the authority to directly merge it. But we suggest each pull request be reviewed by others. A general workflow to submit commit for milvus project is:
1. commiter creates a pull request
2. reviewer review the code and give some suggestions
3. commiter modify the code according to suggestions
4. reviewer accept the pr and give "\lgtm" under the pr
5. maintainer accept the pr and give "\approve" under the pr
6. once "\lgtm" and "\approve" is matched, a robot automatically merge the pr

The proto of milvus-proto is not the latest, but it is compatible to v2.1, just use the milvus-proto as a submodule and sync to #26 is ok.

**Reply by weianweigan** [reply]
@yhmo 
Thanks for your reply!
It's a good workflow.But i don't know who i can request for a review and i need to learn more about github robot.
I need 
```proto
option csharp_namespace = "IO.Milvus.Grpc";
```
in *.proto file so i give up use git submodule now.

**Reply by yhmo** [reply]
> @yhmo Thanks for your reply! It's a good workflow.But i don't know who i can request for a review and i need to learn more about github robot. I need
> 
> ```proto
> option csharp_namespace = "IO.Milvus.Grpc";
> ```
> 
> in *.proto file so i give up use git submodule now.

Ok, I see. I will add this line to milvus-proto later.

**Comment by weianweigan** [comment]
Many thanks!





------------------ Original ------------------
From: groot ***@***.***&gt;
Date: Thu,Aug 4,2022 10:01 AM
To: milvus-io/milvus ***@***.***&gt;
Cc: dududu ***@***.***&gt;, Mention ***@***.***&gt;
Subject: Re: [milvus-io/milvus] about c# sdk (Discussion #18333)





  
@yhmo Thanks for your reply! It's a good workflow.But i don't know who i can request for a review and i need to learn more about github robot. I need
 option csharp_namespace = "IO.Milvus.Grpc";
 
in *.proto file so i give up use git submodule now.
  
Ok, I see. I will add this line to milvus-proto later.

**Reply by yhmo** [reply]
I have merged the pr manually. This is a first step of a sdk project, more tasks need to be done:
- robot for pr check and code merge(I will do this when I have time)
- make milvus-proto to be a submodule
- examples
- unit test cases

**Reply by weianweigan** [reply]
Great!
I will write more unit-tests in the next days.
